dhall-csv/src/Dhall/CsvToDhall.hs

```diff
@@ -1,27 +1,235 @@
+    listConvert :: ExprX -> [Data.Csv.NamedRecord] -> Either CompileError ExprX
+    listConvert (Core.App Core.List recordType@(Core.Record _)) [] = return $ Core.ListLit (Just recordType) Sequence.empty
+    listConvert (Core.App Core.List recordType) [] = Left $ Unsupported recordType
+    listConvert (Core.App Core.List recordType) csv = do
+        a <- traverse (recordConvert recordType) csv
+        return $ Core.ListLit Nothing $ Sequence.fromList a
+    listConvert e _ = Left $ Unsupported e
+
+    recordConvert :: ExprX -> Data.Csv.NamedRecord -> Either CompileError ExprX
+    recordConvert (Core.Record record) csvRecord
+        | extraKeys <- (map decodeUtf8 $ HashMap.keys csvRecord) \\ Map.keys record
+        , strictRecs && not (null extraKeys)
+        = Left $ UnhandledKeys extraKeys
+        | otherwise
+        = do
+            let f = (\k v -> fieldConvert k (Core.recordFieldValue v) (HashMap.lookup (encodeUtf8 k) csvRecord))
+            a <- Map.traverseWithKey f record
+            let a' = Map.mapWithKey (\_ e -> Core.makeRecordField e) a
+            return $ Core.RecordLit a'
+    recordConvert e _ = Left $ Unsupported e
+
+    fieldConvert :: Text -> ExprX -> Maybe Data.Csv.Field -> Either CompileError ExprX
+    -- Unions
+    fieldConvert recordKey t@(Core.Union tm) maybeField = do
+        let f unionKey Nothing =
+                case maybeField of
+                    Nothing -> Left $ MissingKey recordKey
+                    Just field ->
+                        if decodeUtf8 field == unionKey
+                        then Right $ Core.Field t $ Core.makeFieldSelection unionKey
+                        else Left $ Mismatch t field recordKey
+            f unionKey (Just _type) = do
+                expression <- fieldConvert recordKey _type maybeField
+                return (Core.App (Core.Field t $ Core.makeFieldSelection unionKey) expression)
+
+        case (unions, rights (toList (Map.mapWithKey f tm)), maybeField) of
+            (UNone  , _         , _         ) -> Left $ ContainsUnion t
+            (UStrict, xs@(_:_:_), Just field) -> Left $ UndecidableUnion t field xs
+            (UStrict, xs@(_:_:_), Nothing   ) -> Left $ UndecidableMissingUnion t xs
+            (_      , []        , Just field) -> Left $ Mismatch t field recordKey
+            (_      , []        , Nothing   ) -> Left $ MissingKey recordKey
+            (UFirst , x:_       , _         ) -> Right $ x
+            (UStrict, [x]       , _         ) -> Right $ x
+
+    -- Missing Optionals
+    fieldConvert _ (Core.App Core.Optional t) Nothing = return $ Core.App Core.None t
+
+    -- Missing fields
+    fieldConvert key _ Nothing = Left $ MissingKey key
+
+    -- Bools
+    fieldConvert key Core.Bool (Just field) =
+        case readMaybe (init $ tail $ show field) :: Maybe Bool of
+            Nothing -> Left $ Mismatch Core.Bool field key
+            Just v  -> Right $ Core.BoolLit v
```

Comment:
I think this would be simpler (and doesn't require partial functions):

```suggestion
        case field of
            "true"  -> Right (Core.BoolLit True)
            "false" -> Right (Core.BoolLit False)
            _ -> Left $ Mismatch Core.Bool field key
```